test/modules/MetricsValidator/src/ValidateDocumentedMetrics.cs

```diff
@@ -0,0 +1,94 @@
+
+namespace MetricsValidator
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.Edge.Agent.Diagnostics;
+    using Microsoft.Azure.Devices.Edge.Util.Metrics;
+    using Newtonsoft.Json;
+
+    public class ValidateDocumentedMetrics : TestBase
+    {
+        MetricFilter metricFilter;
+
+        public ValidateDocumentedMetrics(TestReporter testReporter, IMetricsScraper scraper, ModuleClient moduleClient)
+            : base(testReporter, scraper, moduleClient)
+        {
+            this.metricFilter = new MetricFilter()
+                .AddTagsToRemove(MetricsConstants.MsTelemetry, MetricsConstants.IotHubLabel, MetricsConstants.DeviceIdLabel);
+        }
+
+        public override async Task Start(CancellationToken cancellationToken)
+        {
+            Console.WriteLine($"Starting {nameof(ValidateDocumentedMetrics)}");
+
+            await this.moduleClient.SendEventAsync(new Message(Encoding.UTF8.GetBytes("Test message to seed metrics")));
+
+            // scrape metrics
+            var metrics = await this.scraper.ScrapeEndpointsAsync(cancellationToken);
+            metrics = this.metricFilter.FilterMetrics(metrics);
+
+            var expected = this.GetExpectedMetrics();
+            HashSet<string> unreturnedMetrics = new HashSet<string>(expected.Keys);
+
+            Console.WriteLine("Got expected metrics");
+            foreach (Metric metric in metrics)
+            {
+                unreturnedMetrics.Remove(metric.Name);
+
+                string metricId = $"{metric.Name}:{JsonConvert.SerializeObject(metric.Tags)}";
+                if (expected.TryGetValue(metric.Name, out string[] expectedTags))
+                {
+                    // Make sure all tags are returned
+                    this.testReporter.Assert(metricId, expectedTags.All(metric.Tags.ContainsKey), $"Expected metric {metric.Name} to contain tags {string.Join(", ", expectedTags)}.\nActual tags: {JsonConvert.SerializeObject(metric.Tags)}");
+                }
+                else
+                {
+                    // Console.WriteLine($"{metricId} not documented");
+                }
+            }
+
+            foreach (string unreturnedMetric in unreturnedMetrics)
+            {
+                this.testReporter.Assert(unreturnedMetric, false, $"Metric did not exist in scrape.");
+            }
+        }
+
+        Dictionary<string, string[]> GetExpectedMetrics()
+        {
+            // read documented metrics
+            IEnumerable<string> lines = Enumerable.Empty<string>();
+            try
+            {
+                var agent = File.ReadAllLines(Path.Combine("doc", "EdgeAgentMetrics.md")).Skip(2);
+                var hub = File.ReadAllLines(Path.Combine("doc", "EdgeHubMetrics.md")).Skip(2);
```

Comment:
Parsing MD files for the list of expected metrics is ok if it's our best option right now. Ultimately, it would be good to consolidate the list of metrics in code somewhere and refer to that instead (since compilers can't verify markdown files).

Reply:
+1 for defining in the code. If the docs move then our tests won't work.

Reply:
There will never be a way for the compiler to verify the metrics. The idea with this it to use the documentation to test, that way we know for sure the documentation is correct and up to date.

Reply:
You can't have the compiler _verify_ the metrics, but if you set up a convention where all metrics are defined in a certain place in a certain way, you could use that. But it isn't an absolute guarantee (to break it, just don't follow the convention...). And I like verifying the docs on some level in the tests.